src/bulk_insert/bulk_insert.c

```diff
@@ -69,181 +78,197 @@ static inline SIValue _BulkInsert_ReadProperty(const char *data, size_t *data_id
+		int rc = _BulkInsert_ProcessFile(gc, data, len, type);
 		UNUSED(rc);
 		ASSERT(rc == BULK_OK);
 	}
-	return BULK_OK;
-}
 
-int _BulkInsert_Insert_Edges(RedisModuleCtx *ctx, GraphContext *gc, int token_count,
-							 RedisModuleString ***argv, int *argc) {
-	int rc;
-	for(int i = 0; i < token_count; i ++) {
-		size_t len;
-		// Retrieve a pointer to the next binary stream and record its length
-		const char *data = RedisModule_StringPtrLen(**argv, &len);
-		*argv += 1;
-		*argc -= 1;
-		rc = _BulkInsert_ProcessRelationFile(ctx, gc, data, len);
-		UNUSED(rc);
-		ASSERT(rc == BULK_OK);
-	}
+	*argv += i;
+	*argc -= i;
+
 	return BULK_OK;
 }
 
-int BulkInsert(RedisModuleCtx *ctx, GraphContext *gc, RedisModuleString **argv, int argc) {
+int BulkInsert(RedisModuleCtx *ctx, GraphContext *gc, RedisModuleString **argv,
+		int argc, uint node_count, uint edge_count) {
+
+	ASSERT(ctx != NULL);
+	ASSERT(gc != NULL);
+	ASSERT(argv != NULL);
 
 	if(argc < 2) {
-		RedisModule_ReplyWithError(ctx, "Bulk insert format error, failed to parse bulk insert sections.");
+		RedisModule_ReplyWithError(ctx, "Bulk insert format error, \
+				failed to parse bulk insert sections.");
 		return BULK_FAIL;
 	}
 
-	// Read the number of node tokens
+	Graph *g = gc->g;
+	int res = BULK_OK;
+
+	// lock graph under write lock
+	// allocate space for new nodes and edges
+	Graph_AcquireWriteLock(g);
+	Graph_AllocateNodes(g, node_count);
+	Graph_AllocateEdges(g, edge_count);
+
+	// read the number of node tokens
 	long long node_token_count;
 	long long relation_token_count;
+
 	if(RedisModule_StringToLongLong(*argv++, &node_token_count)  != REDISMODULE_OK) {
-		RedisModule_ReplyWithError(ctx, "Error parsing number of node descriptor tokens.");
-		return BULK_FAIL;
+		RedisModule_ReplyWithError(ctx, "Error parsing number of node \
+				descriptor tokens.");
+		res = BULK_FAIL;
+		goto cleanup;
 	}
 
+	// read the number of relation tokens
 	if(RedisModule_StringToLongLong(*argv++, &relation_token_count)  != REDISMODULE_OK) {
-		RedisModule_ReplyWithError(ctx, "Error parsing number of relation descriptor tokens.");
-		return BULK_FAIL;
+		RedisModule_ReplyWithError(ctx, "Error parsing number of relation \
+				descriptor tokens.");
+		res = BULK_FAIL;
+		goto cleanup;
 	}
+
 	argc -= 2;
 
 	if(node_token_count > 0) {
-		int rc = _BulkInsert_InsertNodes(ctx, gc, node_token_count, &argv, &argc);
-		if(rc != BULK_OK) {
-			return BULK_FAIL;
-		} else if(argc == 0) {
-			return BULK_OK;
+		// process all node files
+		if(_BulkInsert_ProcessTokens(gc, node_token_count, &argv, &argc,
+				SCHEMA_NODE) != BULK_OK) {
+			res = BULK_FAIL;
+			goto cleanup;
 		}
 	}
 
 	if(relation_token_count > 0) {
-		int rc = _BulkInsert_Insert_Edges(ctx, gc, relation_token_count, &argv, &argc);
-		if(rc != BULK_OK) {
-			return BULK_FAIL;
-		} else if(argc == 0) {
-			return BULK_OK;
+		// Process all relationship files
+		if(_BulkInsert_ProcessTokens(gc, relation_token_count, &argv,
```

Comment:
Seems like `_BulkInsert_ProcessTokens` can not fail

Reply:
That is currently correct, yes.